runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp

```diff
@@ -3360,59 +3359,84 @@ genCAS(TR::Node *node, TR::CodeGenerator *cg, TR_ARM64ScratchRegisterManager *sr
       }
 
    const bool createDoneLabel = (doneLabel == NULL);
-   if (createDoneLabel)
-      {
-      doneLabel = generateLabelSymbol(cg);
-      }
-   /*
-    * Generating the same instruction sequence as __sync_bool_compare_and_swap
-    *
-    * loop:
-    *    ldxrx   resultReg, [addrReg]
-    *    cmpx    resultReg, oldVReg
-    *    bne     done
-    *    stlxrx  resultReg, newVReg, [addrReg]
-    *    cbnz    resultReg, loop
-    *    dmb     ish
-    * done:
-    *    cset    resultReg, eq
-    *
-    */
-   op = is64bit ? TR::InstOpCode::ldxrx : TR::InstOpCode::ldxrw;
-   generateTrg1MemInstruction(cg, op, node, resultReg, new (cg->trHeapMemory()) TR::MemoryReference(addrReg, (int32_t)0, cg));
-   if (oldValueInReg)
-      generateCompareInstruction(cg, node, resultReg, oldVReg, is64bit);
-   else
-      generateCompareImmInstruction(cg, node, resultReg, oldValue, is64bit);
-   if (!createDoneLabel)
-      generateTrg1ImmInstruction(cg, TR::InstOpCode::movzx, node, resultReg, 0); // failure
-   generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_NE);
 
-   if (casWithoutSync)
+   static const bool disableLSE = feGetEnv("TR_aarch64DisableLSE") != NULL;
+   if (comp->target().cpu.supportsFeature(OMR_FEATURE_ARM64_LSE) && (!disableLSE))
       {
-      op = is64bit ? TR::InstOpCode::stxrx : TR::InstOpCode::stxrw;
+      TR_ASSERT_FATAL(oldValueInReg, "Expecting oldValue is in register if LSE is enabled");
+      /*
+       * movx   resultReg, oldVReg
+       * casal  resultReg, newVReg, [addrReg]
+       * cmp    resultReg, oldReg
+       * cset   resultReg, eq
+       */
+      generateMovInstruction(cg, node, resultReg, oldVReg, is64bit);
+      op = casWithoutSync ? (is64bit ? TR::InstOpCode::casx : TR::InstOpCode::casw) : (is64bit ? TR::InstOpCode::casalx : TR::InstOpCode::casalw);
+      generateTrg1MemSrc1Instruction(cg, op, node, resultReg, new (cg->trHeapMemory()) TR::MemoryReference(addrReg, (int32_t)0, cg), newVReg);
+      generateCompareInstruction(cg, node, resultReg, oldVReg, is64bit);
+      generateCSetInstruction(cg, node, resultReg, TR::CC_EQ);
+      if (!createDoneLabel)
+         {
+         generateConditionalBranchInstruction(cg, TR::InstOpCode::b_cond, node, doneLabel, TR::CC_NE);
+         }
       }
    else
       {
-      op = is64bit ? TR::InstOpCode::stlxrx : TR::InstOpCode::stlxrw;
-      }
-   generateTrg1MemSrc1Instruction(cg, op, node, resultReg, new (cg->trHeapMemory()) TR::MemoryReference(addrReg, (int32_t)0, cg), newVReg);
-   generateCompareBranchInstruction(cg, TR::InstOpCode::cbnzx, node, resultReg, loopLabel);
+      TR::LabelSymbol *loopLabel = generateLabelSymbol(cg);
+      generateLabelInstruction(cg, TR::InstOpCode::label, node, loopLabel);
+      if (createDoneLabel)
+         {
+         doneLabel = generateLabelSymbol(cg);
+         }
+      /*
+      * Generating the same instruction sequence as __sync_bool_compare_and_swap
```

Comment:
Please align the position of `*` in this comment.

Reply:
Fixed the indentation of this comment.